src/Service/RefundService.php

```diff
@@ -0,0 +1,333 @@
+        $merchantAccount = $this->configurationService->getMerchantAccount();
+
+        if (!$merchantAccount) {
+            $message = 'No Merchant Account set. ' .
+                'Go to the Adyen plugin configuration panel and finish the required setup.';
+            $this->logger->error($message);
+            throw new AdyenException($message);
+        }
+
+        $pspReference = $orderTransaction->getCustomFields()[PaymentResponseHandler::ORIGINAL_PSP_REFERENCE];
+        $currencyIso = $order->getCurrency()->getIsoCode();
+        $amount = $this->currency->sanitize($order->getAmountTotal(), $currencyIso);
+
+        $params = [
+            'originalReference' => $pspReference,
+            'modificationAmount' => [
+                'value' => $amount,
+                'currency' => $currencyIso
+            ],
+            'merchantAccount' => $merchantAccount
+        ];
+
+        try {
+            $modificationService = new Modification(
+                $this->clientService->getClient($order->getSalesChannelId())
+            );
+
+            return $modificationService->refund($params);
+        } catch (AdyenException $e) {
+            $this->logger->error($e->getMessage());
+            throw $e;
+        }
+    }
+
+    /**
+     * Handle refund notification, by either creating a new adyen_refund entry OR
+     * updating the status of an existing adyen_refund which was initiated on shopware
+     *
+     * @param OrderEntity $order
+     * @param NotificationEntity $notification
+     * @param string $newStatus
+     * @throws AdyenException
+     */
+    public function handleRefundNotification(OrderEntity $order, NotificationEntity $notification, string $newStatus)
+    {
+        $statesToSearch = self::REFUNDABLE_STATES;
+        // This is included for edge cases where an already processed refund, failed (REFUND_FAILED notification)
+        $statesToSearch[] = OrderTransactionStates::STATE_REFUNDED;
+        $orderTransaction = $this->getAdyenOrderTransactionForRefund($order, $statesToSearch);
+
+        $criteria = new Criteria();
+        // Filtering with pspReference since in the future, multiple refunds are possible
+        /** @var RefundEntity $adyenRefund */
+        $criteria->addFilter(new AndFilter([
+            new EqualsFilter('orderTransactionId', $orderTransaction->getId()),
+            new EqualsFilter('pspReference', $notification->getPspreference())
+        ]));
+
+        /** @var RefundEntity $adyenRefund */
+        $adyenRefund = $this->adyenRefundRepository->getRepository()
+            ->search($criteria, Context::createDefaultContext())->first();
+
+        if (is_null($adyenRefund)) {
+            $this->insertAdyenRefund(
+                $order,
+                $notification->getPspreference(),
+                RefundEntity::SOURCE_ADYEN,
+                $newStatus
+            );
+        } else {
+            $this->updateAdyenRefundStatus($adyenRefund, $newStatus);
```

Comment:
Look into if it's possible to use upsert

Reply:
Cannot use this since in the update case, the source would always be overridden and set to `SOURCE_ADYEN`, regardless if it's a new or updated record